{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/lorenzo-arcioni/BLAST-OUT-postprocessing/blob/main/AnnoReport.ipynb)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation Report: Multi-database Annotation Summary\n",
    "\n",
    "This Jupyter notebook is designed to generate a report file based on annotation results obtained from annotation tools such as BLAST or Diamond. To use this notebook, users must have both the annotation results in TSV format and a table containing additional information about the transcripts. By leveraging the power of Jupyter notebooks, users can interactively visualize and analyze their annotation results, generating a comprehensive report that includes detailed statistics and data visualizations. The notebook's intuitive user interface and modular design make it easy to customize the report based on specific research needs. With its ability to quickly and efficiently generate reports from annotation data, this Jupyter notebook is an invaluable tool for researchers working with transcriptomic data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you need to transfer your data from a remote host please use this terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install colab-xterm\n",
    "%load_ext colabxterm\n",
    "%xterm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing libraries\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import numpy as np\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "def get_transcripts_from_id(transcripts, table):\n",
    "    transcripts = transcripts.unique()\n",
    "\n",
    "    dic = dict()\n",
    "\n",
    "    for t in transcripts:\n",
    "        for x in table.transcript:\n",
    "            if re.match(x, t):\n",
    "                dic[t] = x\n",
    "    return dic\n",
    "\n",
    "def make_hyperlink(sseqid, database):\n",
    "    \n",
    "    if database.lower() == 'nr':\n",
    "        protein_accession = sseqid.split(\" \")[0]\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/gene/?term={}\"\n",
    "    else:\n",
    "        protein_accession = sseqid.split(\"|\")[1]\n",
    "        url = \"https://www.uniprot.org/uniprotkb/{}/entry\"\n",
    "        \n",
    "    return '=HYPERLINK(\"%s\", \"%s\")' % (url.format(protein_accession), protein_accession)\n",
    "\n",
    "def get_accession(sseqid, database):\n",
    "\n",
    "    if database.lower() == 'nr':\n",
    "\n",
    "        return sseqid.split(\" \")[0]\n",
    "    else:\n",
    "        \n",
    "        return sseqid.split(\"|\")[1]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, the user must customize the generation parameters (by appropriately modifying the variables) following the instructions in the comments in the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert the names (or paths) of the tsv files\n",
    "files = [\n",
    "    \"blastp/blastp_nr.tsv\",\n",
    "    \"blastp/blastp_tr.tsv\",\n",
    "    \"blastp/blastp_sp.tsv\"\n",
    "] \n",
    "\n",
    "# Insert the titles of the graph\n",
    "title = \"Blastp\"\n",
    "\n",
    "# Insert the databases names (the order must match the result files order)\n",
    "databases_names =[\n",
    "    \"Nr\", \n",
    "    \"TrEMBL\",\n",
    "    \"Swiss-Prot\",\n",
    "]\n",
    "\n",
    "# Insert the table (with additional informations) path\n",
    "table_path = \"./blastp/tables/\"\n",
    "\n",
    "# Insert the path of the report\n",
    "path = \"./\" + title + \"_\" + table_path.split(\"/\")[-1]\n",
    "\n",
    "# Set the outformat\n",
    "# e.g. \n",
    "# outfmt = \"qseqid qlen sseqid sallseqid slen qstart qend sstart send qseq full_qseq sseq full_sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos qframe btop cigar staxids sscinames sskingdoms skingdoms sphylums stitle salltitles qcovhsp scovhsp qtitle qqual full_qqual qstrand\"\n",
    "# If there are column names in the file then set outfmt = None\n",
    "#outfmt = \"qseqid qlen sseqid sallseqid slen qstart qend sstart send qseq full_qseq sseq full_sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos qframe btop cigar stitle salltitles qcovhsp scovhsp qtitle qqual full_qqual qstrand\"\n",
    "outfmt = None\n",
    "# Columns names (modify this list by inserting the column names of the report)\n",
    "features = [\"transcript\", \"row\", \"log2FoldChange\", \"padj\", \n",
    "            \"protein_accession\", \"sequence_identity\", \"alignment_length\", \n",
    "            \"evalue\", \"database\", \"gene\", \"locus_name\", \"sequence_description\",\n",
    "            \"sequence_length\", \"organism\", \"protein_product\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "table = pd.read_csv(table_path, sep='\\t')\n",
    "table.index.name = 'transcript'\n",
    "\n",
    "table.reset_index(inplace=True)\n",
    "\n",
    "for i in range(len(files)):\n",
    "\n",
    "    #Import the dataset\n",
    "    if outfmt == None:\n",
    "        df_tmp = pd.read_csv(files[i], sep=\"\\t\")\n",
    "    else:\n",
    "        df_tmp = pd.read_csv(files[i], sep=\"\\t\", names=outfmt.split())\n",
    "\n",
    "    df_tmp['qseqid'] = df_tmp['qseqid'].map(get_transcripts_from_id(df_tmp['qseqid'], table))\n",
    "    df_tmp['database'] = databases_names[i]\n",
    "    df_tmp['row'] = title\n",
    "    # df_tmp['sequence_identity'] = df_tmp.pident\n",
    "    # df_tmp['alignment_length'] = df_tmp.length\n",
    "    # df_tmp['evalue'] = df_tmp.evalue\n",
    "    # df_tmp['sequence_description'] = df_tmp.stitle\n",
    "    # df_tmp['sequence_length'] = df_tmp.slen\n",
    "\n",
    "    df_tmp.rename(columns={'pident': 'sequence_identity',\n",
    "                           'length': 'alignment_length',\n",
    "                           'stitle': 'sequence_description',\n",
    "                           'slen': 'sequence_length'\n",
    "    }, inplace=True)\n",
    "\n",
    "    if \"OS=\" not in df_tmp.sequence_description[0]:    \n",
    "        def get_sciname(x):\n",
    "            \n",
    "            try:\n",
    "                os_index = - x[::-1].index('[')\n",
    "            except:\n",
    "                return \"\"\n",
    "\n",
    "            return x[os_index:-1]\n",
    "\n",
    "        # Useful functions\n",
    "        def get_protein_function(x):\n",
    "\n",
    "            x_l = x.split(\" \")\n",
    "\n",
    "            try:\n",
    "                nex = ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('[')))])\n",
    "            except:\n",
    "                return \"\"\n",
    "\n",
    "            return nex\n",
    "        \n",
    "        def get_locus_name(x):\n",
    "            return None\n",
    "        \n",
    "        def get_gene(x):\n",
    "            return None\n",
    "    else:\n",
    "        def get_sciname(x):\n",
    "\n",
    "            os_index = x.index('OS=')\n",
    "            ox_index = x.index('OX=')\n",
    "\n",
    "            return x[os_index+3:ox_index-1]\n",
    "\n",
    "        # Useful functions\n",
    "        def get_protein_function(x):\n",
    "\n",
    "            x_l = x.split(\" \")\n",
    "\n",
    "            return ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('OS=')))])\n",
    "        \n",
    "        def get_locus_name(x):\n",
    "            return x.split(\"|\")[2]\n",
    "        \n",
    "        def get_gene(x):\n",
    "\n",
    "            try:\n",
    "                gn_index = x.index('GN=')\n",
    "                pe_index = x.index('PE=')\n",
    "            except:\n",
    "                return None\n",
    "            return x[gn_index+3:pe_index-1]\n",
    "        \n",
    "    df_tmp['gene'] = df_tmp.sequence_description.apply(lambda x: get_gene(x))\n",
    "    df_tmp['organism'] = df_tmp.sequence_description.apply(lambda x: get_sciname(x))\n",
    "    df_tmp['protein_accession'] = df_tmp.apply(lambda x: get_accession(x.sseqid, x.database), axis=1)\n",
    "    df_tmp['protein_product'] = df_tmp.sequence_description.apply(lambda x: get_protein_function(x))\n",
    "    df_tmp['locus_name'] = df_tmp.sseqid.apply(lambda x: get_locus_name(x))\n",
    "\n",
    "    df_tmp = pd.merge(df_tmp, table, left_on='qseqid', right_on='transcript', how='left')\n",
    "\n",
    "    df = pd.concat([df, df_tmp[features]])\n",
    "\n",
    "df.sort_values(['transcript', 'evalue'], inplace=True)\n",
    "\n",
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>row</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>sequence_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>database</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_name</th>\n",
       "      <th>sequence_description</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>organism</th>\n",
       "      <th>protein_product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GKNQ01001237.1</td>\n",
       "      <td>Blastp</td>\n",
       "      <td>-1.121591</td>\n",
       "      <td>0.018273</td>\n",
       "      <td>VDM39180.1</td>\n",
       "      <td>81.1</td>\n",
       "      <td>928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VDM39180.1 unnamed protein product [Toxocara c...</td>\n",
       "      <td>936</td>\n",
       "      <td>Toxocara canis</td>\n",
       "      <td>unnamed protein product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GKNQ01001237.1</td>\n",
       "      <td>Blastp</td>\n",
       "      <td>-1.121591</td>\n",
       "      <td>0.018273</td>\n",
       "      <td>TMS38063.1</td>\n",
       "      <td>72.6</td>\n",
       "      <td>916</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TMS38063.1 hypothetical protein L596_004871 [S...</td>\n",
       "      <td>939</td>\n",
       "      <td>Steinernema carpocapsae</td>\n",
       "      <td>hypothetical protein L596_004871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GKNQ01001237.1</td>\n",
       "      <td>Blastp</td>\n",
       "      <td>-1.121591</td>\n",
       "      <td>0.018273</td>\n",
       "      <td>CDJ94054.1</td>\n",
       "      <td>70.9</td>\n",
       "      <td>929</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CDJ94054.1 Alanine dehydrogenase PNT and Sacch...</td>\n",
       "      <td>931</td>\n",
       "      <td>Haemonchus contortus</td>\n",
       "      <td>Alanine dehydrogenase PNT and Saccharopine deh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GKNQ01001237.1</td>\n",
       "      <td>Blastp</td>\n",
       "      <td>-1.121591</td>\n",
       "      <td>0.018273</td>\n",
       "      <td>VDO37093.1</td>\n",
       "      <td>69.3</td>\n",
       "      <td>948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>VDO37093.1 unnamed protein product [Haemonchus...</td>\n",
       "      <td>951</td>\n",
       "      <td>Haemonchus placei</td>\n",
       "      <td>unnamed protein product</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GKNQ01001237.1</td>\n",
       "      <td>Blastp</td>\n",
       "      <td>-1.121591</td>\n",
       "      <td>0.018273</td>\n",
       "      <td>EYC41722.1</td>\n",
       "      <td>70.5</td>\n",
       "      <td>914</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>EYC41722.1 hypothetical protein Y032_0558g3408...</td>\n",
       "      <td>932</td>\n",
       "      <td>Ancylostoma ceylanicum</td>\n",
       "      <td>hypothetical protein Y032_0558g3408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript     row  log2FoldChange      padj protein_accession  \\\n",
       "0  GKNQ01001237.1  Blastp       -1.121591  0.018273        VDM39180.1   \n",
       "1  GKNQ01001237.1  Blastp       -1.121591  0.018273        TMS38063.1   \n",
       "2  GKNQ01001237.1  Blastp       -1.121591  0.018273        CDJ94054.1   \n",
       "3  GKNQ01001237.1  Blastp       -1.121591  0.018273        VDO37093.1   \n",
       "4  GKNQ01001237.1  Blastp       -1.121591  0.018273        EYC41722.1   \n",
       "\n",
       "   sequence_identity  alignment_length  evalue database  gene locus_name  \\\n",
       "0               81.1               928     0.0       Nr  None       None   \n",
       "1               72.6               916     0.0       Nr  None       None   \n",
       "2               70.9               929     0.0       Nr  None       None   \n",
       "3               69.3               948     0.0       Nr  None       None   \n",
       "4               70.5               914     0.0       Nr  None       None   \n",
       "\n",
       "                                sequence_description  sequence_length  \\\n",
       "0  VDM39180.1 unnamed protein product [Toxocara c...              936   \n",
       "1  TMS38063.1 hypothetical protein L596_004871 [S...              939   \n",
       "2  CDJ94054.1 Alanine dehydrogenase PNT and Sacch...              931   \n",
       "3  VDO37093.1 unnamed protein product [Haemonchus...              951   \n",
       "4  EYC41722.1 hypothetical protein Y032_0558g3408...              932   \n",
       "\n",
       "                  organism                                    protein_product  \n",
       "0           Toxocara canis                            unnamed protein product  \n",
       "1  Steinernema carpocapsae                   hypothetical protein L596_004871  \n",
       "2     Haemonchus contortus  Alanine dehydrogenase PNT and Saccharopine deh...  \n",
       "3        Haemonchus placei                            unnamed protein product  \n",
       "4   Ancylostoma ceylanicum                hypothetical protein Y032_0558g3408  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to remove repetitive information (e.g. trasncript name ecc..) run the cell below (you can also modify given list)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.duplicated(subset=['transcript', 'row', 'log2FoldChange', 'padj']), 'transcript':'padj'] = ''\n",
    "unmatched_transcript = pd.Series(list(set(table.transcript) - set(df.transcript)), \n",
    "                                 name='unmatched_transcript')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the cell below to save the report in tsv format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(path + '.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the cell below to save the report in xlsx format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openpyxl.utils import get_column_letter\n",
    "\n",
    "df_writer = pd.ExcelWriter(path + '.xlsx') \n",
    "\n",
    "# Scrivi il dataframe nel foglio di lavoro\n",
    "df.to_excel(df_writer, sheet_name='report', index=False)\n",
    "\n",
    "# Ottieni l'oggetto Worksheet dall'oggetto ExcelWriter\n",
    "worksheet = df_writer.sheets['report']\n",
    "\n",
    "# Imposta la larghezza delle colonne in modo automatico\n",
    "for column in range(df.shape[1]):\n",
    "    column_length = max(df.iloc[:, column].astype(str).map(len).max(), len(df.columns[column])) + 3\n",
    "    column_letter = get_column_letter(column + 1)\n",
    "    worksheet.column_dimensions[column_letter].width = column_length\n",
    "\n",
    "# Salva il foglio di lavoro\n",
    "df_writer.close()\n",
    "\n",
    "# Calcolo i trascritti che non hano mappato\n",
    "df_writer = pd.ExcelWriter(path + \"_unmatch\" + \".xlsx\")\n",
    "unmatched_transcript.to_excel(df_writer, sheet_name='unmatched transcripts', index=False)\n",
    "\n",
    "# Ottieni l'oggetto Worksheet dall'oggetto ExcelWriter\n",
    "worksheet = df_writer.sheets['unmatched transcripts']\n",
    "\n",
    "# Imposta la larghezza delle colonne in modo automatico\n",
    "worksheet.column_dimensions['A'].width = max(unmatched_transcript.astype(str).map(len).max(), len(unmatched_transcript.name)) + 3\n",
    "\n",
    "# Salva il foglio di lavoro\n",
    "df_writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
