{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing libraries\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import numpy as np\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert the names (or paths) of the tsv files\n",
    "files = [\n",
    "    \"bombina/bombina_corset_DEGS__unk_not-unk.fasta.transdecoder.pep_nr.tsv\",\n",
    "    \"bombina/bombina_corset_DEGS__unk_not-unk.fasta.transdecoder.pep_tr.tsv\",\n",
    "    \"bombina/bombina_corset_DEGS__unk_not-unk.fasta.transdecoder.pep_sp.tsv\"   \n",
    "]\n",
    "\n",
    "# Insert the titles of the graph\n",
    "title = \"bombina_pachypus_blastp\"\n",
    "\n",
    "# Insert the databases names\n",
    "databases_names =[\n",
    "    \"Nr\", \n",
    "    \"TrEMBL\",\n",
    "    \"Swiss-Prot\",\n",
    "]\n",
    "\n",
    "table_path = \"./bombina/bombina_unref_vs_not_unkref_table_padj_0.05----log2fc_1 (4).tsv\"\n",
    "\n",
    "# Insert the path of the graphs\n",
    "path = \"bombina/report/\" + title\n",
    "\n",
    "# Set the outformat\n",
    "# e.g. \n",
    "# outfmt = \"qseqid qlen sseqid sallseqid slen qstart qend sstart send qseq full_qseq sseq full_sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos qframe btop cigar staxids sscinames sskingdoms skingdoms sphylums stitle salltitles qcovhsp scovhsp qtitle qqual full_qqual qstrand\"\n",
    "# If there are column names in the file then set outfmt = None\n",
    "outfmt = \"qseqid qlen sseqid sallseqid slen qstart qend sstart send qseq full_qseq sseq full_sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos qframe btop cigar staxids sscinames sskingdoms skingdoms sphylums stitle salltitles qcovhsp scovhsp qtitle qqual full_qqual qstrand\"\n",
    "\n",
    "# Columns names\n",
    "features = [\"transcript\", \"row\", \"log2FoldChange\", \"padj\", \n",
    "            \"protein_accession\", \"sequence_identity\", \"alignment_length\", \n",
    "            \"evalue\", \"database\", \"gene\", \"locus_name\", \"sequence_description\",\n",
    "            \"sequence_length\", \"organism\", \"protein_product\"]\n",
    "\n",
    "table = pd.read_csv(table_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_100481_length_1662_cov_45.942320_g39073_i0</td>\n",
       "      <td>113.572774</td>\n",
       "      <td>3.729238</td>\n",
       "      <td>0.749879</td>\n",
       "      <td>4.973122</td>\n",
       "      <td>6.588303e-07</td>\n",
       "      <td>0.001008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_100687_length_1658_cov_35.724701_g39182_i0</td>\n",
       "      <td>34.441283</td>\n",
       "      <td>2.713316</td>\n",
       "      <td>0.640121</td>\n",
       "      <td>4.238757</td>\n",
       "      <td>2.247607e-05</td>\n",
       "      <td>0.016458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_100697_length_1658_cov_14.098680_g3928_i1</td>\n",
       "      <td>20.244586</td>\n",
       "      <td>-5.263078</td>\n",
       "      <td>1.052069</td>\n",
       "      <td>-5.002597</td>\n",
       "      <td>5.656316e-07</td>\n",
       "      <td>0.000894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_101001_length_1653_cov_22.559899_g39352_i0</td>\n",
       "      <td>34.277619</td>\n",
       "      <td>2.555678</td>\n",
       "      <td>0.589651</td>\n",
       "      <td>4.334222</td>\n",
       "      <td>1.462765e-05</td>\n",
       "      <td>0.012317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_101191_length_1650_cov_9.495262_g39455_i0</td>\n",
       "      <td>19.673311</td>\n",
       "      <td>-5.888866</td>\n",
       "      <td>1.467984</td>\n",
       "      <td>-4.011534</td>\n",
       "      <td>6.032561e-05</td>\n",
       "      <td>0.036529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        transcript    baseMean  \\\n",
       "0  NODE_100481_length_1662_cov_45.942320_g39073_i0  113.572774   \n",
       "1  NODE_100687_length_1658_cov_35.724701_g39182_i0   34.441283   \n",
       "2   NODE_100697_length_1658_cov_14.098680_g3928_i1   20.244586   \n",
       "3  NODE_101001_length_1653_cov_22.559899_g39352_i0   34.277619   \n",
       "4   NODE_101191_length_1650_cov_9.495262_g39455_i0   19.673311   \n",
       "\n",
       "   log2FoldChange     lfcSE      stat        pvalue      padj  \n",
       "0        3.729238  0.749879  4.973122  6.588303e-07  0.001008  \n",
       "1        2.713316  0.640121  4.238757  2.247607e-05  0.016458  \n",
       "2       -5.263078  1.052069 -5.002597  5.656316e-07  0.000894  \n",
       "3        2.555678  0.589651  4.334222  1.462765e-05  0.012317  \n",
       "4       -5.888866  1.467984 -4.011534  6.032561e-05  0.036529  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript(row):\n",
    "    for t in table.transcript:\n",
    "        if re.match(t, row):\n",
    "            return t\n",
    "    return None\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i in range(len(files)):\n",
    "\n",
    "    #Import the dataset\n",
    "    df_tmp = pd.read_csv(files[i], sep=\"\\t\", names=outfmt.split())\n",
    "\n",
    "    df_tmp['transcript'] = df_tmp['qseqid'].apply(get_transcript)\n",
    "    df_tmp['row'] = title\n",
    "    df_tmp['sequence_identity'] = df_tmp.pident\n",
    "    df_tmp['alignment_length'] = df_tmp.length\n",
    "    df_tmp['evalue'] = df_tmp.evalue\n",
    "    df_tmp['sequence_description'] = df_tmp.stitle\n",
    "    df_tmp['sequence_length'] = df_tmp.slen\n",
    "    df_tmp['database'] = databases_names[i]\n",
    "\n",
    "    if \"OS=\" not in df_tmp.stitle[0]:    \n",
    "        def get_sciname(x):\n",
    "            \n",
    "            os_index = - x[::-1].index('[')\n",
    "\n",
    "            return x[os_index:-1]\n",
    "\n",
    "        # Useful functions\n",
    "        def get_protein_function(x):\n",
    "\n",
    "            x_l = x.split(\" \")\n",
    "\n",
    "            return ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('[')))])\n",
    "        \n",
    "        def get_protein_accession(x):\n",
    "            return x.split(\" \")[0]\n",
    "        \n",
    "        def get_locus_name(x):\n",
    "            return None\n",
    "        \n",
    "        def get_gene(x):\n",
    "            return None\n",
    "    else:\n",
    "        def get_sciname(x):\n",
    "\n",
    "            os_index = x.index('OS=')\n",
    "            ox_index = x.index('OX=')\n",
    "\n",
    "            return x[os_index+3:ox_index-1]\n",
    "\n",
    "        # Useful functions\n",
    "        def get_protein_function(x):\n",
    "\n",
    "            x_l = x.split(\" \")\n",
    "\n",
    "            return ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('OS=')))])\n",
    "        \n",
    "        def get_protein_accession(x):\n",
    "            return x.split(\"|\")[1]\n",
    "        \n",
    "        def get_locus_name(x):\n",
    "            return x.split(\"|\")[2]\n",
    "        \n",
    "        def get_gene(x):\n",
    "\n",
    "            try:\n",
    "                gn_index = x.index('GN=')\n",
    "                pe_index = x.index('PE=')\n",
    "            except:\n",
    "                return None\n",
    "            return x[gn_index+3:pe_index-1]\n",
    "        \n",
    "    df_tmp['gene'] = df_tmp.stitle.apply(lambda x: get_gene(x))\n",
    "    df_tmp['organism'] = df_tmp.stitle.apply(lambda x: get_sciname(x))\n",
    "    df_tmp['protein_accession'] = df_tmp.sseqid.apply(lambda x: get_protein_accession(x))\n",
    "    df_tmp['protein_product'] = df_tmp.stitle.apply(lambda x: get_protein_function(x))\n",
    "    df_tmp['locus_name'] = df_tmp.sseqid.apply(lambda x: get_locus_name(x))\n",
    "\n",
    "    df_tmp = pd.merge(df_tmp, table, on='transcript')\n",
    "\n",
    "    df = pd.concat([df, df_tmp[features]])\n",
    "\n",
    "df.sort_values(['transcript', 'evalue'], inplace=True)\n",
    "\n",
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.duplicated(subset=['transcript', 'row', 'log2FoldChange', 'padj']), 'transcript':'padj'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>row</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>padj</th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>sequence_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>database</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_name</th>\n",
       "      <th>sequence_description</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>organism</th>\n",
       "      <th>protein_product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_100481_length_1662_cov_45.942320_g39073_i0</td>\n",
       "      <td>bombina_pachypus_blastp</td>\n",
       "      <td>3.729238</td>\n",
       "      <td>0.001008</td>\n",
       "      <td>A0A6P7WME3</td>\n",
       "      <td>38.9</td>\n",
       "      <td>332</td>\n",
       "      <td>1.870000e-64</td>\n",
       "      <td>TrEMBL</td>\n",
       "      <td>LOC115458754</td>\n",
       "      <td>A0A6P7WME3_9AMPH</td>\n",
       "      <td>tr|A0A6P7WME3|A0A6P7WME3_9AMPH epidermal diffe...</td>\n",
       "      <td>340</td>\n",
       "      <td>Microcaecilia unicolor</td>\n",
       "      <td>epidermal differentiation-specific protein-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XP_030044427.1</td>\n",
       "      <td>38.9</td>\n",
       "      <td>332</td>\n",
       "      <td>4.810000e-64</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_030044427.1 epidermal differentiation-speci...</td>\n",
       "      <td>340</td>\n",
       "      <td>Microcaecilia unicolor</td>\n",
       "      <td>epidermal differentiation-specific protein-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>A0A6P8NSY5</td>\n",
       "      <td>38.9</td>\n",
       "      <td>337</td>\n",
       "      <td>1.020000e-63</td>\n",
       "      <td>TrEMBL</td>\n",
       "      <td>LOC117349897</td>\n",
       "      <td>A0A6P8NSY5_GEOSA</td>\n",
       "      <td>tr|A0A6P8NSY5|A0A6P8NSY5_GEOSA epidermal diffe...</td>\n",
       "      <td>339</td>\n",
       "      <td>Geotrypetes seraphini</td>\n",
       "      <td>epidermal differentiation-specific protein-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XP_033779482.1</td>\n",
       "      <td>38.9</td>\n",
       "      <td>337</td>\n",
       "      <td>2.620000e-63</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_033779482.1 epidermal differentiation-speci...</td>\n",
       "      <td>339</td>\n",
       "      <td>Geotrypetes seraphini</td>\n",
       "      <td>epidermal differentiation-specific protein-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XP_028822286.1</td>\n",
       "      <td>37.5</td>\n",
       "      <td>339</td>\n",
       "      <td>7.370000e-63</td>\n",
       "      <td>Nr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>XP_028822286.1 epidermal differentiation-speci...</td>\n",
       "      <td>339</td>\n",
       "      <td>Denticeps clupeoides</td>\n",
       "      <td>epidermal differentiation-specific protein-like</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        transcript                      row  \\\n",
       "0  NODE_100481_length_1662_cov_45.942320_g39073_i0  bombina_pachypus_blastp   \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4                                                                             \n",
       "\n",
       "  log2FoldChange      padj protein_accession  sequence_identity  \\\n",
       "0       3.729238  0.001008        A0A6P7WME3               38.9   \n",
       "1                             XP_030044427.1               38.9   \n",
       "2                                 A0A6P8NSY5               38.9   \n",
       "3                             XP_033779482.1               38.9   \n",
       "4                             XP_028822286.1               37.5   \n",
       "\n",
       "   alignment_length        evalue database          gene        locus_name  \\\n",
       "0               332  1.870000e-64   TrEMBL  LOC115458754  A0A6P7WME3_9AMPH   \n",
       "1               332  4.810000e-64       Nr          None              None   \n",
       "2               337  1.020000e-63   TrEMBL  LOC117349897  A0A6P8NSY5_GEOSA   \n",
       "3               337  2.620000e-63       Nr          None              None   \n",
       "4               339  7.370000e-63       Nr          None              None   \n",
       "\n",
       "                                sequence_description  sequence_length  \\\n",
       "0  tr|A0A6P7WME3|A0A6P7WME3_9AMPH epidermal diffe...              340   \n",
       "1  XP_030044427.1 epidermal differentiation-speci...              340   \n",
       "2  tr|A0A6P8NSY5|A0A6P8NSY5_GEOSA epidermal diffe...              339   \n",
       "3  XP_033779482.1 epidermal differentiation-speci...              339   \n",
       "4  XP_028822286.1 epidermal differentiation-speci...              339   \n",
       "\n",
       "                 organism                                  protein_product  \n",
       "0  Microcaecilia unicolor  epidermal differentiation-specific protein-like  \n",
       "1  Microcaecilia unicolor  epidermal differentiation-specific protein-like  \n",
       "2   Geotrypetes seraphini  epidermal differentiation-specific protein-like  \n",
       "3   Geotrypetes seraphini  epidermal differentiation-specific protein-like  \n",
       "4    Denticeps clupeoides  epidermal differentiation-specific protein-like  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "arrays = [\n",
    "\n",
    "    [\"bar\", \"bar\", \"baz\", \"baz\", \"foo\", \"foo\", \"qux\", \"qux\"],\n",
    "\n",
    "    [\"one\", \"two\", \"one\", \"two\", \"one\", \"two\", \"one\", \"two\"],\n",
    "\n",
    "]\n",
    "\n",
    "index = pd.MultiIndex.from_tuples(list(zip(*arrays)), names=[\"first\", \"second\"])\n",
    "\n",
    "pd.DataFrame(np.random.randn(8), index=index).to_csv(\"prova.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'index' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 9\u001b[0m\n\u001b[1;32m      6\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(f, sep\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m\"\u001b[39m, names\u001b[39m=\u001b[39moutfmt\u001b[39m.\u001b[39msplit())\n\u001b[1;32m      8\u001b[0m \u001b[39m#Drop Nan rows\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m df\u001b[39m.\u001b[39mdropna(subset\u001b[39m=\u001b[39m[index], inplace\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m     11\u001b[0m \u001b[39m#Create and append the dict\u001b[39;00m\n\u001b[1;32m     12\u001b[0m sets_list\u001b[39m.\u001b[39mappend(\u001b[39mset\u001b[39m(df[index]) \u001b[39m-\u001b[39m {index})\n",
      "\u001b[0;31mNameError\u001b[0m: name 'index' is not defined"
     ]
    }
   ],
   "source": [
    "#Creating a dictonary list where each dict contains all sequences hits count\n",
    "sets_list = []\n",
    "\n",
    "for f in files:\n",
    "    #Import the dataset\n",
    "    df = pd.read_csv(f, sep=\"\\t\", names=outfmt.split())\n",
    "    \n",
    "    #Drop Nan rows\n",
    "    df.dropna(subset=[index], inplace=True)\n",
    "    \n",
    "    #Create and append the dict\n",
    "    sets_list.append(set(df[index]) - {index})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
