{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e5860491",
   "metadata": {},
   "source": [
    "<a href=\"https://colab.research.google.com/github/lorenzo-arcioni/BioSwiftAnalytics/blob/main/notebooks/AnnoViz.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>\n",
    "\n",
    "# Visualizing Insights: A Jupyter Notebook for Enhanced Annotation Result Interpretation\n",
    "\n",
    "This Jupyter Notebook support the analysis of homology-based annotation results generated by BLAST or Diamond annotation software. The tool allow users to generate informative graphs and reports to aid in the interpretation of tabular annotation files obtained during the annotation process. The required dependencies for the notebook to run include the pandas, matplotlib, numpy, ipywidgets and math libraries.\n",
    "\n",
    "By following the instructions provided, users will be guided through the process of setting the parameters for their analysis. Once these parameters have been set, users can execute the notebook's cells one by one. It is strongly recommended that users follow the instructions step-by-step to ensure proper use of the tool.\n",
    "\n",
    "Through interactive code snippets and visualizations, users will learn how to manipulate and plot data generated from homology-based annotation pipelines, including sequence alignments metadata. With this Jupyter Notebook, users can quickly and easily explore their annotation data and gain deeper insights into the underlying biological processes at play.\n",
    "\n",
    "First of all, exec the cell below to import all the necessary libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ac9d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing libraries\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\t\n",
    "\n",
    "from math import log10, floor\n",
    "from IPython.display import Markdown\n",
    "\n",
    "import matplotlib.cm as cm\n",
    "import ipywidgets as widgets\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "#Find the exponent of a float number in scientific notation\n",
    "def find_exp(number) -> int:\n",
    "    if abs(number) == 0:\n",
    "        return 0\n",
    "    base10 = log10(abs(number))\n",
    "    return abs(floor(base10))\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 110"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b3c03e65",
   "metadata": {},
   "source": [
    "## Preliminary phase"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3512d1b8",
   "metadata": {},
   "source": [
    "### Parameter configuration section\n",
    "\n",
    "This section is responsible for loading the dataframe and allows the user to set the following parameters for their analysis:\n",
    "- Input file name\n",
    "- BLAST output format (each subsequent section will require specific information)\n",
    "- Title for their graphs (we recommend a title that reflects the annotation information, such as the database name or BLAST type, or the scientific name of the organism being analyzed)\n",
    "- Path for saving the image or graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "753e58ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Insert the name (or path) of the tsv file\n",
    "filename = \"../data/culex_pipiens-longest_orf-blastp-nr.tsv\"\n",
    "\n",
    "#Insert the outformat\n",
    "# If the column names are in the first line of the input file then leave it as None \n",
    "#outformat = None\n",
    "outfmt = \"qseqid qlen sseqid slen evalue bitscore score length pident stitle\"\n",
    "\n",
    "#Insert the title of graph\n",
    "title = \"Culex pipiens BlastP NR\"\n",
    "\n",
    "# Insert the output file path between the quotes\n",
    "path = \"../results/\" + title.lower().replace(\" \", \"_\")\n",
    "\n",
    "# If you want to save the plot data also as a tsv file, set save_data to True\n",
    "save_data = False\n",
    "\n",
    "#Import the dataset (don't modify anything here)\n",
    "if outfmt is None:\n",
    "    df = pd.read_csv(filename, sep=\"\\t\")\n",
    "else:\n",
    "    df = pd.read_csv(filename, sep=\"\\t\", names=outfmt.split())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "20c08cf2",
   "metadata": {},
   "source": [
    "### Deleting less significant hits\n",
    "\n",
    "This cell allows the user to obtain only the most significant hit for each sequence by eliminating all other hits from the dataframe. It is recommended to execute this cell only if the user wants to remove all non-significant hits and keep only the most significant hit for each sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55dc875",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.groupby('qseqid').first().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa54d8ea",
   "metadata": {},
   "source": [
    "To have an idea of how appear the loaded data you can run the cell below!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92b0fb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "43f3bab5",
   "metadata": {},
   "source": [
    "## Section for analyzing the hit rate of protein functions\n",
    "\n",
    "The hitrate graph is important in homology annotation because it provides a measure of the quality of the annotation. The hitrate represents the percentage of sequences that have a significant hit in the database for a particular organism. If the hitrate is low, it may indicate that the database is incomplete or that the annotation algorithm is not performing well for that particular organism. Conversely, a high hitrate suggests a good coverage in the database and a reliable annotation. Therefore, the hitrate graph can help researchers assess the quality of the annotation and identify areas where improvements are needed.\n",
    "\n",
    "In this section, it is possible to obtain bar charts highlighting the most common protein functions that were found across all input sequences. \n",
    "\n",
    "Two charts will be generated: \n",
    "- one showing the protein ID bar chart and \n",
    "- one showing the protein function name count bar chart.\n",
    "\n",
    "Please note that to run these cells, it is necessary to have \"sseqid\" and \"stitle\" as part of the output format information."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bd385c15",
   "metadata": {},
   "source": [
    "#### Accession Graph"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "abea1ef5",
   "metadata": {},
   "source": [
    "A graph of accessions is a graphical representation that shows the number of hits associated with each protein accession in a BLAST or Diamond search result. It can be used to quickly identify the most frequently matched proteins in the search and to gain insights into the functional annotation of the input sequences. The Y-axis of the graph represents the protein accession numbers, while the X-axis represents the number of hits associated with each accession.\n",
    "\n",
    "Now, run this cell to calculate the accession rate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a639737",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Series with accession id and count\n",
    "hits = df['sseqid'].value_counts()\n",
    "\n",
    "# Insert the number of results (default 15)\n",
    "results_number = 15\n",
    "\n",
    "# Drop the last len(h) - results_number lines, only if  len(h) > results_number\n",
    "if len(hits) > results_number:\n",
    "    hits.drop(hits.tail(len(hits) - results_number).index, inplace=True)\n",
    "\n",
    "# Set appropriate figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (8, min(results_number//2 - results_number//4 + 1, len(hits)//2 - len(hits)//4 + 2))\n",
    "\n",
    "# Sortting and adjusting data\n",
    "hits.sort_values(inplace=True)\n",
    "hits.reset_index(drop=True);"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9169c914",
   "metadata": {},
   "source": [
    "Run this cell to plot the results!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49f8de90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating the graph with color\n",
    "colors = cm.Blues(hits.values / float(max(hits.values)))\n",
    "plt.barh(hits.index, hits.values, color=colors)\n",
    "plt.ylim([-1,len(hits.index)])\n",
    "plt.xlim([0, hits.iloc[-1] + hits.iloc[0]//5])\n",
    "\n",
    "# Insert labels\n",
    "for index, value in enumerate(hits.values):\n",
    "    plt.text(value, index, \" \" + str(value), va='center')\n",
    "\n",
    "# Set title and axis labels\n",
    "plt.xlabel('Number of hits')\n",
    "plt.ylabel('Accession')\n",
    "plt.title(title + ' protein accessions distribution')\n",
    "\n",
    "# Save and plot!\n",
    "if save_data:\n",
    "    hits.to_csv(path + '_protein_accessions.tsv', sep=\"\\t\")\n",
    "    \n",
    "plt.savefig(path + '_protein_accessions.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07f47312",
   "metadata": {},
   "source": [
    "#### Section for selecting annotation database"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "78ee5230",
   "metadata": {},
   "source": [
    "In this section, you can select the database on which the annotation operations have been performed. This is because the result storage system varies from database to database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a97963",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Widget radio button\n",
    "database = widgets.Select(\n",
    "    options=['NR', 'TrEMBL', 'Swiss-Prot'],\n",
    "    value='NR',\n",
    "    description='Select the database:',\n",
    "    disabled=False,\n",
    "    style={'description_width': 'initial'}\n",
    ")\n",
    "display(database)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c828bebe",
   "metadata": {},
   "source": [
    "Now execute this cell to make effective the changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52e05cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if database.value == 'NR':    \n",
    "    def get_sciname(x):\n",
    "        \n",
    "        os_index = - x[::-1].index('[')\n",
    "\n",
    "        return x[os_index:-1]\n",
    "\n",
    "    # Useful functions\n",
    "    def get_protein_function(x):\n",
    "\n",
    "        x_l = x.split(\" \")\n",
    "\n",
    "        return ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('[')))])\n",
    "else:\n",
    "    def get_sciname(x):\n",
    "\n",
    "        os_index = x.index('OS=')\n",
    "        ox_index = x.index('OX=')\n",
    "\n",
    "        return x[os_index+3:ox_index-1]\n",
    "\n",
    "    # Useful functions\n",
    "    def get_protein_function(x):\n",
    "\n",
    "        x_l = x.split(\" \")\n",
    "\n",
    "        return ' '.join(x_l[1:x_l.index(next(x for x in x_l if x.startswith('OS=')))])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a3df7460",
   "metadata": {},
   "source": [
    "### Proteins graph"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7e97998c",
   "metadata": {},
   "source": [
    "In this section, a bar chart can be generated showing the count of protein function names associated with the annotation.\n",
    "You can change the default values of the results number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cef2c739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new column in the dataframe\n",
    "df['protein'] = df['stitle'].apply(lambda x: get_protein_function(x))\n",
    "\n",
    "# Series with accession id and count\n",
    "proteins  = df['protein'].value_counts()\n",
    "\n",
    "# Insert all index to remove from the plot\n",
    "try:\n",
    "    proteins = proteins[~proteins.index.str.contains('ncharacterized', regex=True, na=False)]\n",
    "    proteins = proteins[~proteins.index.str.contains('ncharacterised', regex=True, na=False)]\n",
    "    proteins = proteins[~proteins.index.str.contains('ypothetical', regex=True, na=False)]\n",
    "    proteins = proteins[~proteins.index.str.contains('nnamed', regex=True, na=False)]\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# Insert the number of results\n",
    "results_number = 15\n",
    "\n",
    "# Drop the last len(h) - results_number lines, only if  len(h) > results_number\n",
    "if len(proteins) > results_number:\n",
    "    proteins.drop(proteins.tail(len(proteins) - results_number).index, inplace=True)\n",
    "\n",
    "# Sorting and adjusting data\n",
    "proteins.sort_values(inplace=True)\n",
    "proteins.reset_index(drop=True)\n",
    "proteins.sort_values(ascending=True, inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5813fcf7",
   "metadata": {},
   "source": [
    "Now execute cell below to plot the bar chart!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0990732",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a horizontal bar chart\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.barh(proteins.index, proteins.values, color='#855F5F')\n",
    "\n",
    "#Set appropriate figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (8, min(results_number//2 - results_number//4 + 1, len(proteins)//2 - len(proteins)//4 + 2))\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "ax.set_xlim([0, proteins.iloc[-1] + proteins.iloc[0]])\n",
    "ax.set_ylim([-1, len(proteins.index)])\n",
    "\n",
    "# Insert labels with values inside each bar\n",
    "for index, value in enumerate(proteins.values):\n",
    "    ax.text(value, index, \" \" + str(value), ha='left', va='center')\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('Number of proteins', fontsize=12)\n",
    "ax.set_ylabel('Protein', fontsize=12)\n",
    "ax.set_title(title + ' protein distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    proteins.to_csv(path + '_protein_distribution.tsv', sep=\"\\t\")\n",
    "plt.savefig(path + '_protein_distribution.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9a93a939",
   "metadata": {},
   "source": [
    "## Section for analyzing the hit rate of organisms\n",
    "\n",
    "Similarly to the protein function chart, we can obtain a bar chart that allows us to understand whether the organism under examination is evolutionarily close to another organism present in the database.\n",
    "\n",
    "In this section, it is possible to obtain bar charts highlighting the most common organisms that were found across all input sequences.\n",
    "Please note that to run these cells, it is necessary to have \"sseqid\" and \"stitle\" as part of the output format information."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "92dcdbe7",
   "metadata": {},
   "source": [
    "#### Scientific names graph"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b91cd399",
   "metadata": {},
   "source": [
    "You can change the default values of the results number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52319b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new column in the dataframe\n",
    "df['scientific_name']  = df['stitle'].apply(lambda x: get_sciname(x))\n",
    "\n",
    "# Series with scientific names\n",
    "sci_names = df['scientific_name'].value_counts()\n",
    "\n",
    "# Insert the number of results (default 15)\n",
    "results_number = 15\n",
    "\n",
    "if len(sci_names) > results_number:\n",
    "    sci_names.drop(sci_names.tail(len(sci_names) - results_number).index, inplace=True)\n",
    "\n",
    "# Sorting and adjusting data\n",
    "sci_names.sort_values(inplace=True)\n",
    "sci_names.reset_index(drop=True);"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "79542e9e",
   "metadata": {},
   "source": [
    "Now execute the cell below to plot!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbe77a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a horizontal bar chart\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.barh(sci_names.index, sci_names.values, color='#69b3a2')\n",
    "\n",
    "#Set appropriate figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (8, min(results_number//2 - results_number//4 + 1, len(sci_names)//2 - len(sci_names)//4 + 2))\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "ax.set_xlim([0, sci_names.max() + sci_names.max()*0.1])\n",
    "ax.set_ylim([-1, len(sci_names.index)])\n",
    "\n",
    "# Insert labels with values inside each bar\n",
    "for index, value in enumerate(sci_names.values):\n",
    "    ax.text(value, index, \" \" + str(value), ha='left', va='center')\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('Number of hits', fontsize=12)\n",
    "ax.set_ylabel('Scientific Name', fontsize=12)\n",
    "ax.set_title(title + ' species distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    sci_names.to_csv(path + '_species_distribution.tsv', sep='\\t')\n",
    "plt.savefig(path + '_species_distribution.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c8c5a329",
   "metadata": {},
   "source": [
    "## Hit count bar chart\n",
    "\n",
    "In this section, it is possible to generate a bar chart of hits for each match in the annotation results (if the limit has been set [here](#deleting-less-significant-hits), it will be a bar chart with all values at 1). \n",
    "\n",
    "Here it is also possible to modify the default value of the number of results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad8c3c79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Series with names and count\n",
    "hits = df['qseqid'].value_counts()\n",
    "\n",
    "# Insert the number of results (default 30)\n",
    "results_number = 30\n",
    "\n",
    "# Drop the last len(h) - results_number lines, only if  len(h) > results_number\n",
    "if len(hits) > results_number:\n",
    "    hits.drop(hits.tail(len(hits) - results_number).index, inplace=True)\n",
    "\n",
    "# Set appropriate figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (8, min(results_number//2 - results_number//4 + 1, len(hits)//2 - len(hits)//4 + 2))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a6b0fbde",
   "metadata": {},
   "source": [
    "And now plot!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "578a22a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a horizontal bar chart\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "hits.sort_values(inplace=True)\n",
    "\n",
    "ax.barh(hits.index, hits.values, color='#69b3a2')\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "ax.set_xlim([0, hits.max() + hits.max()*0.1])\n",
    "ax.set_ylim([-1, len(hits.index)])\n",
    "\n",
    "# Insert labels with values inside each bar\n",
    "for index, value in enumerate(hits.values):\n",
    "    ax.text(value, index, \" \" + str(value), ha='left', va='center')\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('Number of hits', fontsize=12)\n",
    "ax.set_ylabel('Sequences', fontsize=12)\n",
    "ax.set_title(title + ' hits distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    hits.to_csv(path + '_hits_distribution.tsv', sep='\\t')\n",
    "plt.savefig(path + '_hits_distribution.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c6dbbd84",
   "metadata": {},
   "source": [
    "## Distribution of e-value from annotation results file\n",
    "\n",
    "The e-value is a statistical measure used in bioinformatics to assess the significance of sequence similarity search results. It represents the expected number of hits with a similar score that would be obtained by chance alone when searching a database of a certain size. A lower e-value indicates a more significant match between the query sequence and the database sequence.\n",
    "\n",
    "In this section, you can extract the e-value from the annotation results file and plot a histogram representing the distribution of the e-value for all the hits. To perform this analysis, it is necessary to have the 'evalue' parameter in the tabular output format.\n",
    "\n",
    "Now, run the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4efc9840",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Series with e-value of subject sequenceid\n",
    "evalues_series = df[\"evalue\"].copy()\n",
    "\n",
    "#Setting the figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f3e6bd61",
   "metadata": {},
   "source": [
    "And now you can plot!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b61cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort the DataFrame by e-value\n",
    "evalues_series.sort_values(inplace=True, ascending=True)\n",
    "\n",
    "# Create a list of exponents of all e-value of dataframe\n",
    "evalues = list(evalues_series.values)\n",
    "evalues = [np.format_float_scientific(x, precision=4, exp_digits=3) for x in evalues]\n",
    "evalues_exp = [find_exp(float(x)) for x in evalues]\n",
    "\n",
    "# Create the histogram with a color palette\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "n, bins, patches = ax.hist(evalues_exp, bins=[x for x in range(0, max(evalues_exp), max(evalues_exp)//40)], \n",
    "                           edgecolor='black', linewidth=0.5, alpha=0.7)\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "#ax.set_xlim([-40, 1])\n",
    "#ax.set_ylim([0, max(ax.get_yticks())])\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('E-value (1e-x)', fontsize=12)\n",
    "ax.set_ylabel('Hits', fontsize=12)\n",
    "ax.set_title(title + ' E-value distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    evalues_series.to_csv(path + '_evalues.tsv', sep='\\t')\n",
    "plt.savefig(path + '_evalue.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2adca986",
   "metadata": {},
   "source": [
    "## Bit-score distribution analysis\n",
    "\n",
    "The bitscore is a statistical measure that represents the relative quality of the alignment between a query sequence and a subject sequence in a sequence database. It is calculated from the raw score, which is the sum of the match, mismatch, and gap scores between the query and subject sequences, and takes into account the size of the database being searched. A higher bitscore indicates a better alignment and a higher degree of similarity between the query and subject sequences. The bitscore is commonly used as a criterion for evaluating the significance of sequence alignments and identifying potential homologous sequences.\n",
    "\n",
    "In this section, similarly to the previous one, it is possible to generate a histogram graph that shows the distribution of the bitscore value for each match.\n",
    "\n",
    "To perform this analysis, it is necessary to have the 'bitscore' parameter in the tabular output format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83415fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Series with bit-score of subject sequenceid\n",
    "bitscores = df[\"bitscore\"].copy()\n",
    "\n",
    "#Setting the figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e18ce515",
   "metadata": {},
   "source": [
    "And now plot!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38e54863",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sorting by bit-score\n",
    "bitscores.sort_values(inplace=True, ascending=True)\n",
    "\n",
    "# Adjusting the indexes\n",
    "bitscores.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# Create the histogram\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "n, bins, patches = ax.hist(bitscores.values, bins=[x for x in range(0, int(max(bitscores.values)), int(max(bitscores.values))//50)], color='steelblue', edgecolor='black')\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "ax.set_xlim([0, max(bitscores.values)])\n",
    "ax.set_ylim([0, max(ax.get_yticks())])\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('Bit-score', fontsize=12)\n",
    "ax.set_ylabel('Hits', fontsize=12)\n",
    "ax.set_title(title + ' Bit-score distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Calculate mean and standard deviation\n",
    "mean = bitscores.values.mean()\n",
    "std = bitscores.values.std()\n",
    "\n",
    "# Add mean and standard deviation lines\n",
    "ax.axvline(mean, color='red', linestyle='--', linewidth=2)\n",
    "ax.axvline(mean - std, color='orange', linestyle='--', linewidth=2)\n",
    "ax.axvline(mean + std, color='orange', linestyle='--', linewidth=2)\n",
    "\n",
    "# Adding the legend\n",
    "plt.legend(['Mean', 'Standard deviation'])\n",
    "\n",
    "# Set the color of each patch based on its height\n",
    "bin_centers = 0.5 * (bins[:-1] + bins[1:])\n",
    "colormap = plt.cm.ScalarMappable(cmap='Blues')\n",
    "colormap.set_clim(0, max(n))\n",
    "for c, p in zip(colormap.to_rgba(n), patches):\n",
    "    plt.setp(p, 'facecolor', c)\n",
    "\n",
    "# Zoom in around the mean\n",
    "ax.set_xlim([mean - 2*std, mean + 10*std])\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    bitscores.to_csv(path + '_bitscore.tsv', sep='\\t')\n",
    "plt.savefig(path +'_bitscore.png', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1470e8f9",
   "metadata": {},
   "source": [
    "## Length and Similarity distribution"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f7cb53a0",
   "metadata": {},
   "source": [
    "In a homology annotation process, the similarity score and length score are important measures to assess the degree of similarity between a query sequence and a sequence in a reference database.\n",
    "\n",
    "The similarity score is typically calculated as the percentage of identical amino acids or nucleotides between the two sequences, and it provides a measure of how closely related the two sequences are. A high similarity score suggests that the two sequences share a common evolutionary origin and may have similar biological functions.\n",
    "\n",
    "The length score, on the other hand, measures the length of the matching region between the two sequences. It is important because it provides additional information about the degree of similarity between the two sequences. A longer matching region indicates a stronger similarity between the two sequences, while a shorter region may suggest that the two sequences are only partially homologous.\n",
    "\n",
    "Overall, the similarity and length scores are important measures to consider when assessing the significance of a homology match between a query sequence and a sequence in a reference database.\n",
    "\n",
    "To obtain the results in this section, it is essential to have length ('length') and percent identity ('pident') information in the output file."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "04ab88de",
   "metadata": {},
   "source": [
    "### Length distribution"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "684065b7",
   "metadata": {},
   "source": [
    "This chart displays the distribution of all calculated sequence lengths, shows the overall performance of the alignments and helps to adjust the annotation score in the annotation step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "372a33f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Series with length of subject sequenceid\n",
    "length = df[\"length\"].copy()\n",
    "\n",
    "#Setting the figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "92a034ca",
   "metadata": {},
   "source": [
    "And now plot!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4214e6f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sorting by bit-score\n",
    "length.sort_values(inplace=True, ascending=True)\n",
    "\n",
    "# Adjusting the indexes\n",
    "length.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# Create the histogram\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "n, bins, patches = ax.hist(length.values, bins=[x for x in range(0, int(max(length.values)), int(max(length.values))//50)], color='steelblue', edgecolor='black')\n",
    "\n",
    "# Set the x-axis and y-axis limits\n",
    "ax.set_xlim([0, max(length.values)])\n",
    "ax.set_ylim([0, max(ax.get_yticks())])\n",
    "\n",
    "# Set axis labels and title\n",
    "ax.set_xlabel('Sequence-length', fontsize=12)\n",
    "ax.set_ylabel('Hits', fontsize=12)\n",
    "ax.set_title(title + ' Length distribution', fontsize=14)\n",
    "\n",
    "# Add grid lines\n",
    "ax.xaxis.grid(color='gray', linestyle='--', alpha=0.7)\n",
    "\n",
    "# Calculate mean and standard deviation\n",
    "mean = length.values.mean()\n",
    "std  = length.values.std()\n",
    "\n",
    "# Add mean and standard deviation lines\n",
    "ax.axvline(mean, color='red', linestyle='--', linewidth=2)\n",
    "ax.axvline(mean - std, color='orange', linestyle='--', linewidth=2)\n",
    "ax.axvline(mean + std, color='orange', linestyle='--', linewidth=2)\n",
    "\n",
    "# Adding the legend\n",
    "plt.legend(['Mean', 'Standard deviation'])\n",
    "\n",
    "# Set the color of each patch based on its height\n",
    "bin_centers = 0.5 * (bins[:-1] + bins[1:])\n",
    "colormap = plt.cm.ScalarMappable(cmap='Blues')\n",
    "colormap.set_clim(0, max(n))\n",
    "for c, p in zip(colormap.to_rgba(n), patches):\n",
    "    plt.setp(p, 'facecolor', c)\n",
    "\n",
    "# Zoom in around the mean\n",
    "ax.set_xlim([mean - 2*std, mean + 10*std])\n",
    "\n",
    "# Save and display the chart\n",
    "if save_data:\n",
    "    length.to_csv(path + '_lendistribution.tsv', sep='\\t')\n",
    "plt.savefig(path +'_lendistribution.png', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "01bde3e3",
   "metadata": {},
   "source": [
    "### Similarity distribution"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f6ad55a1",
   "metadata": {},
   "source": [
    "This chart displays the distribution of all calculated sequence similarities (percentages), shows the overall performance of the alignments and helps to adjust the annotation score in the annotation step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3caa186",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the identity percentage\n",
    "pidents = df[\"pident\"].copy()\n",
    "\n",
    "#Setting the figsize\n",
    "plt.rcParams[\"figure.figsize\"] = (5, 5)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fb90511b",
   "metadata": {},
   "source": [
    "And now plot the results!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd50a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the figure size\n",
    "plt.figure(figsize=(8, 6))\n",
    "\n",
    "# Sort the data and create the histogram\n",
    "pidents.sort_values(inplace=True, ascending=True)\n",
    "pidents.reset_index(drop=True, inplace=True)\n",
    "plt.hist(pidents.values, bins=[x for x in range(0, max(int(max(pidents.values)), 1), max(1, int(max(pidents.values))//50))])\n",
    "\n",
    "# Add labels and title\n",
    "plt.xlabel('Sequence Similarity (%)')\n",
    "plt.ylabel('Number of Hits')\n",
    "plt.title(title + ' Similarity Distribution')\n",
    "\n",
    "# Add a grid\n",
    "plt.grid(True)\n",
    "\n",
    "# Save and show the plot\n",
    "if save_data:\n",
    "    pidents.to_csv(path + '_similarity.tsv', sep='\\t')\n",
    "plt.savefig(path + '_similarity.png', bbox_inches = 'tight', dpi=300)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
